New Serverless Pattern - Route53-ALB-S3 (hosting a private website) (#2438)
aws-samples/serverless-patterns: route53-alb-s3/README.md

> # Internal static website hosting using Amazon Route53, ALB and S3

Reviewer: ```suggestion
# Internal static website hosting using Amazon Route53, Application Load Balancer and Amazon S3
```

## Deployment Instructions

1. From the command line, use AWS SAM to deploy the AWS resources for the pattern as specified in the template.yml file:
    ```
    sam deploy --guided
    ```
2. During the prompts:
    * Enter a stack name
    * Enter the desired AWS Region e.g. us-east-1 
    * Enter PrivateWebsiteDomainName e.g. hello-world.example.com
    * Allow SAM CLI to create IAM roles with the required permissions
    * Leave all other options to default values

    Once you have run `sam deploy --guided` mode once and saved arguments to a configuration file (samconfig.toml), you can use `sam deploy` in future to use these defaults.

3. Note the outputs from the SAM deployment process. These contain the resource names and/or ARNs which are used for testing.

4. From the command line, run the below command to upload the `index.html` to the S3 bucket. Replace `{WebsiteBucket}` with the corresponding output value of the `sam deploy` command.
    ```
    aws s3 cp index.html s3://{WebsiteBucket}
    ```

## How it works

Please refer to the architecture diagram below:

![End to End Architecture](architecture.png)

> This template deploys Amazon Route53, Application Load Balancer (ALB), VPC Endpoint and Amazon S3. The private website URL will be exposed via Route 53 private hosted zone. This URL is accessible from within VPC.

Reviewer: ```suggestion
This template deploys a Route53 hosted zone, an Application Load Balancer (ALB), a VPC Endpoint and a S3 bucket. The private website URL will be exposed via the Route 53 private hosted zone. This URL is accessible from within the VPC.
```

## Testing

1. [Create a CloudShell VPC environment](https://docs.aws.amazon.com/cloudshell/latest/userguide/creating-vpc-environment.html). Enter the following values:
    * Name - enter your preferred environment name.
    * VPC - `VPCId` from the `sam deploy` output.
    * Subnet - Select any of the two available subnets under the above VPC.
    * Security - `ALBSecurityGroup` from the `sam deploy` output.

2. From the CloudShell command prompt run the following command to access your private website. Replace `{PrivateWebsiteUrl}` with the corresponding output value of the `sam deploy` command:
    ```

>     curl -v -k -L {PrivateWebsiteUrl}   

Reviewer: please give a one sentence explanation what these parameters do. Also, is -v necessary?

Reviewer 2: @bfreiberg I have added some explanation to the command. Yeah `-v` is probably not needed and have removed it. I was using it primarily to test if the cert chain was indeed a private one or not.

    ```
3. You can also test by spinning up an EC2 instance within the same VPC or from another location that has private connectivity to the VPC. However, if you try to access the URL from outside network, it will not be reachable.

## Cleanup
 

> 1. Delete the contents of the S3 bucket.

Reviewer: ```suggestion
1. Delete the contents of the S3 bucket. Make sure to use the correct bucket name.
```
